### PR TITLE
[xef-evaluator] - Improvements

### PR DESCRIPTION
**Pull request by tomascayuelas**
Improvements:
- Content validation
- Richer domain
- Change DSL
- Adding tests

Don't hesitate to comment anything.

### CODE REVIEW

File: evaluator-example/src/main/kotlin/com/xebia/functional/xef/evaluator/examples/TestExample.kt
Comment (javipacheco) [inline evaluator-example/src/main/kotlin/com/xebia/functional/xef/evaluator/examples/TestExample.kt:31]: :clap:

**Review by javipacheco — APPROVED** [review]
🚀  Great job!!

I have tested the evaluator locally and everything works as expected.

**Review by fedefernandez — COMMENTED** [review]
This looks great @tomascayuelas. Just one comment, how hard would be to use the `kotlin.test` framework instead `kotest`? In Arrow they followed the approach of using kotest just for the assertions due to compatibility reasons with the platform.

**Review by fedefernandez — DISMISSED** [review]

**Comment by tomascayuelas** [comment]
> This looks great @tomascayuelas. Just one comment, how hard would be to use the `kotlin.test` framework instead `kotest`? In Arrow they followed the approach of using kotest just for the assertions due to compatibility reasons with the platform.

Thanks @fedefernandez I'll check it and see if it's simple.

**Comment by fedefernandez** [comment]
@tomascayuelas I made a migration in this commit, in case it helps
https://github.com/xebia-functional/xef/pull/561/commits/2c5a8528d71345f8db8514377bc0575befec98f0
Arrow's closed related PRs are also a good source of info

File: evaluator/src/main/kotlin/com/xebia/functional/xef/evaluator/EitherOps.kt
Comment (fedefernandez) [inline evaluator/src/main/kotlin/com/xebia/functional/xef/evaluator/EitherOps.kt:11]: Question, in Kotlin is this a good practice (nested the functions in an object)? I have seen in the project the extension functions defined at top level

File: evaluator/build.gradle.kts
Comment (fedefernandez) [inline evaluator/build.gradle.kts:36]: Bad intent?

File: evaluator/src/main/kotlin/com/xebia/functional/xef/evaluator/models/errors/SuiteSpecError.kt
Comment (fedefernandez) [inline evaluator/src/main/kotlin/com/xebia/functional/xef/evaluator/models/errors/SuiteSpecError.kt:3]: Should this file be called `ValidationError` instead?

File: evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/EitherOpsSpec.kt
Comment (fedefernandez) [inline evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/EitherOpsSpec.kt:24]: Each time this test is executed, it runs once with random data, right? In that case, I think it would be better to use fixed values because, right now, can lead to flaky tests. Not a big deal though, just a preference.

File: evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/ItemSpecBuilderSpec.kt
Comment (fedefernandez) [inline evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/ItemSpecBuilderSpec.kt:149]: All the invalid tests could be replaced by a property test that generates a tuple of the invalid `ItemSpec` + the error list.

File: evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/SuiteBuilderSpec.kt
Comment (fedefernandez) [inline evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/SuiteBuilderSpec.kt:203]: Same as above. I prefer the way `ItemSpecBuilderSpec` defines the data, but here we can also group some tests in a single one.

File: evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/utils/Generators.kt
Comment (fedefernandez) [inline evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/utils/Generators.kt:11]: I think this could be probably more efficient to generate a random-size list of one of the blank spaces

File: evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/TestModelsSpec.kt
Comment (fedefernandez) [inline evaluator/src/test/kotlin/com/xebia/functional/xef/evaluator/TestModelsSpec.kt:17]: These can me coded as a property-based tests

**Review by fedefernandez — COMMENTED** [review]
Left some minor comments. I think it would be worthy to address or discuss

**Comment by tomascayuelas** [comment]
Thank you very much @fedefernandez for your comments/review.
As we have already spoken by slack, I take care of them, I think they are very good, and can be better code. Thank you!

File: evaluator/src/main/kotlin/com/xebia/functional/xef/evaluator/models/errors/SuiteSpecError.kt
Comment (tomascayuelas) [reply evaluator/src/main/kotlin/com/xebia/functional/xef/evaluator/models/errors/SuiteSpecError.kt:3]: I moved the ValidationError to a new file.

**Review by nomisRev — CHANGES_REQUESTED** [review]
We're here using Kotest inside Kotlin Test 🤕 Never seen that before 😅 Why are we using both?

We can also define a simple DSL by ourselves that covers the requirements of this PR. It's probably not much more than what we've already defined on top of Kotest.

Additionally, I agree with @fedefernandez that we should either properly do prop testing (with random/configurable SEED), or use fixed data.

A couple small questions, thoughts, and concerns on some the domain classes defined in this PR.

**Comment by fedefernandez** [comment]
> @fedefernandez @tomascayuelas is this still relevant?

@raulraja I feel it's. I can take a look and prepare for reviewing

> We're here using Kotest inside Kotlin Test 🤕 Never seen that before 😅 Why are we using both?

We faced some binary issues and found the issues you created in Arrow. We're using the kotlin test framework but kotest for the assertions.

https://github.com/arrow-kt/arrow/issues/3183

**Comment by nomisRev** [comment]
I see I am blocking this PR @fedefernandez.

To give some context on Arrow, that was to avoid rewriting more code and to political continue supporting Kotest. We replaced the Kotest runner only because we wanted to support WASM, and be able to build EAP releases (as part of the community builds).

However, I don't want to block this PR on that matter. I do feel we can simplify some things here, Junit also support "property based testing" in the form of table based testing. Not sure if Qwik works for KMP, but I've used it for kotlin-test on JVM.

That leaves the same concerns as you originally raised @fedefernandez concerning testing practices. I am removing my request for changes however to unblock this PR from my side.

**Comment by fedefernandez** [comment]
Thanks @nomisRev. I have already started resolving some comments; I'll get back to you once I have something.

**Comment by fedefernandez** [comment]
I'm afraid I don't have enough context to fix the conflicts. Therefore, I vote to close the PR.

**Pull request closed without merging**
